File: web/src/ponto.js
/**
 * Ponto JavaScript implementation for WebViews
 *
 * @see  http://github.com/wikia-apps/Ponto
 *
 * @author Federico "Lox" Lucignano <federico@wikia-inc.com>
 *
 */

/*global define, require, PontoUriProtocol*/
(function (context) {
	'use strict';

	/**
	 * AMD support
	 *
	 * @private
	 *
	 * @type {Boolean}
	 */
	var amd = false;

	/**
	 * module constructor
	 *
	 * @private
	 *
	 * @return {Object} The module reference
	 */
	function ponto() {
		var
			/**
			 * [Constant] Name of the URI protocol optionally
			 * registered by the native layer
			 *
			 * @private
			 *
			 * @type {String}
			 *
			 * @see  protocol.request
			 * @see  protocol.response
			 */
				PROTOCOL_NAME = 'ponto',

			/**
			 * [Constant] Represents a completed request
			 *
			 * @private
			 *
			 * @type {Number}
			 *
			 * @see  Ponto.acceptResponse
			 */
				RESPONSE_COMPLETE = 0,

			/**
			 * [Constant] Represents a failed request with errors
			 *
			 * @private
			 *
			 * @type {Number}
			 *
			 * @see  Ponto.acceptResponse
			 */
				RESPONSE_ERROR = 1,

			/**
			 * [Constant] Indicates that the target is a native platform
			 *
			 * @type {Number}
			 */
				TARGET_NATIVE = 0,

			/**
			 * [Constant] Indicates that the target is an iframe
			 *
			 * @type {Number}
			 */
				TARGET_IFRAME = 1,

			/**
			 * [Constant] Indicates that the target is an iframe parent window
			 *
			 * @type {Number}
			 */
				TARGET_IFRAME_PARENT = 2,

			/**
			 * Window to communicate with (if iframe transport is overriden)
			 *
			 * @type {Window}
			 */
				targetWindow,

			/**
			 * Request / Response dispatcher
			 *
			 * @type {PontoDispatcher}
			 */
				dispatcher  = new PontoDispatcher(context),

			/**
			 * Registry for complete/error callbacks
			 *
			 * @private
			 *
			 * @type {Object}
			 */
				callbacks = {},

			/**
			 * Protocol helper
			 * registered by native platforms that
			 * have the capability to do so (e.g. Android) or
			 * implemented directly for those that don't (e.g iOS)
			 *
			 * @type {Object}
			 */
				protocol = nativeProtocol(),

				exports;

		/**
		 * Returns a valid communication protocol for native platform
		 * @returns {*|{request: request, response: response}}
		 */
		function nativeProtocol () {
			return context.PontoProtocol || {
				//the only other chance is for the native layer to register
				//a custom protocol for communicating with the webview (e.g. iOS)
				request: function (execContext, target, method, params, callbackId) {
					if (execContext.location && execContext.location.href) {
						execContext.location.href = PROTOCOL_NAME + ':///request?target=' + encodeURIComponent(target) +
							'&method=' + encodeURIComponent(method) +
							((Object.keys(params).length) ? '&params=' + encodeURIComponent(JSON.stringify(params)) : '') +
							((callbackId) ? '&callbackId=' + encodeURIComponent(callbackId) : '');
					} else {
						throw new Error('Context doesn\'t support User Agent location API');
					}
				},
				response: function (execContext, callbackId, params) {
					if (execContext.location && execContext.location.href) {
						execContext.location.href = PROTOCOL_NAME + ':///response?callbackId=' + encodeURIComponent(callbackId) +
							((params) ? '&params=' + encodeURIComponent(JSON.stringify(params)) : '');
					} else {
						throw new Error('Context doesn\'t support User Agent location API');
					}
				}
			};
		}

		/**
		 * Returns a valid communication protocol for the iframe
		 * @returns {{request: request, response: response}}
		 */
		function iframeProtocol () {
			return {
				request: function (execContext, target, method, params, callbackId, async) {
					if (targetWindow.postMessage) {
						targetWindow.postMessage({
							protocol: PROTOCOL_NAME,
							action: 'request',
							target: target,
							method: method,
							params: params,
							async: async,
							callbackId: callbackId
						}, targetWindow.location.origin);
					} else {
						throw new Error('Target context does not support postMessage');
					}
				},
				response: function (execContext, callbackId, result) {
					if (targetWindow.postMessage) {
						targetWindow.postMessage({
							protocol: PROTOCOL_NAME,
							action: 'response',
							type: (result && result.type) ? result.type : RESPONSE_COMPLETE,
							params: result,
							callbackId: callbackId
						}, targetWindow.location.origin);
					} else {
						throw new Error('Target context does not support postMessage');
					}
				}
			};
		}

		/**
		 * 'message' Event handler
		 * @param {Event} event
		 */
		function onMessage(event){
			if (event.data && event.data.protocol === PROTOCOL_NAME) {
				dispatcher[event.data.action](event.data);
			}
		}


		/**
		 * Request handler base class constructor
		 *
		 * @public
		 */
		function PontoBaseHandler() {}

		/**
		 * Request handler factory method, each subclass of
		 * PontoBaseHandler needs to implement this static method
		 */
		PontoBaseHandler.getInstance = function () {
			throw new Error('The getInstance method needs to be overridden in PontoBaseHandler subclasses');
		};

		/**
		 * PontoBaseHandler extension pattern
		 *
		 * @param {PontoBaseHandler} constructor The reference to a constructor
		 * of a type to be converted in a PontoBaseHandler subtype
		 *
		 * @example
		 * function MyHandler(){...}
		 * Ponto.PontoBaseHandler.derive(MyHandler);
		 * MyHandler.getInstance = function() {...};
		 */
		PontoBaseHandler.derive = function (constructor) {
			constructor.handlerSuper = PontoBaseHandler;
			constructor.getInstance = PontoBaseHandler.getInstance;
			constructor.prototype = new PontoBaseHandler();
		};

		/**
		 * Dispatches a request sent by the native layer
		 *
		 * @private
		 *
		 * @param {Object} scope The execution scope
		 * @param {Mixed} target A reference to a constructor or a static instance
		 * @param {RequestParams} data An hash containing the parameters associated to the request
		 */
		function dispatchRequest(scope, target, data) {
			var instance,
				result;

			if (target && target.handlerSuper === PontoBaseHandler && target.getInstance) {
				instance = target.getInstance();

				//unfortunately we need to instantiate before being able to
				if (instance[data.method]) {
					if (data.async) {
						instance[data.method](data.params, data.callbackId);
					} else {
						result = instance[data.method](data.params);

						if (data.callbackId && protocol && protocol.response) {
							protocol.response(scope, data.callbackId, result);
						}
					}
				}
			}
		}

		/**
		 * Function called by the native layer when answering a request
		 *
		 * @public
		 *
		 * @param {Object} scope The execution scope
		 * @param {ResponseParams} data An hash containing the parameters associated to the response
		 */
		function dispatchResponse(data) {
			var
				callbackId = data.callbackId,
				cbGroup = callbacks[callbackId],
				callback,
				responseType;

			if (cbGroup) {
				responseType = data.type;

				switch (responseType) {
					case RESPONSE_COMPLETE:
						callback = cbGroup.complete;
						break;
					case RESPONSE_ERROR:
					default:
						callback = cbGroup.error;
						break;
				}

				if (callback) {
					callback(data.params);
				}

				delete callbacks[callbackId];
			}
		}

		/**
		 * Overrides the protocol target (default: native)
		 * @param {Number} _target
		 * @param {Number | undefined}_targetWindow
		 * provide if target is an iframe
		 */
		function setTarget(_target, _targetWindow) {
			switch (_target) {
				case TARGET_IFRAME:
					if (_targetWindow.top && _targetWindow !== _targetWindow.top) {
						targetWindow = _targetWindow;
					} else {
						throw new Error('Bad iframe content window provided.');
					}
					break;
				case TARGET_IFRAME_PARENT:
					if (context.top && context.top !== context) {
						targetWindow = context.top;
					} else {
						throw new Error('No possible communication in this context.');
					}
					break;
				default:
					protocol = nativeProtocol();
					if (PontoDispatcher.prototype.respond) {
						delete PontoDispatcher.prototype.respond;
					}
					return;
			}
			protocol = iframeProtocol();

			context.addEventListener('message', onMessage, false);

			/**
			 * Enables manual trigger of the response when async operation needed
			 * @param {Object} result response params, optionally with 'type' field
			 * @param {Number} callbackId
			 */
			PontoDispatcher.prototype.respond = function (result, callbackId) {
				protocol.response(this.context, callbackId, result);
			};
		}

		/**
		 * RequestParams constructor
		 *
		 * Extracts and normalizes the parameters out of a JSON-encoded string
		 * passed in by a request from the native context
		 *
		 * @private
		 *
		 * @param {String} data JSON-encoded string describing parameters
		 * for a request to Ponto
		 */
		function RequestParams(data) {
			var hash = typeof data === 'string' ? JSON.parse(data) : data;

			this.target = hash.target;
			this.method = hash.method;
			this.params = hash.params;
			this.callbackId = hash.callbackId;
			this.async = hash.async;
		}

		/**
		 * ResponseParams constructor
		 *
		 * Extracts and normalizes the parameters out of a JSON-encoded string
		 * passed in by a response from the native context
		 *
		 * @private
		 *
		 * @param {String} data JSON-encoded string describing parameters
		 * for a response from Ponto
		 */
		function ResponseParams(data) {
			var hash = typeof data === 'string' ? JSON.parse(data) : data;

			this.type = parseInt(hash.type, 10);
			this.callbackId = hash.callbackId;
			this.params = hash.params;
		}

		/**
		 * Ponto dispatcher constructor
		 *
		 * @public
		 *
		 * @param {Object} dispatchContext The execution scope to bind to this instance
		 */
		function PontoDispatcher(dispatchContext) {
			this.context = dispatchContext;
		}

		/**
		 * Sets the context/scope to bind to the PontoDispatcher instance
		 *
		 * @public
		 *
		 * @param {Object} scope The context/scope to bind to the instance
		 */
		PontoDispatcher.prototype.setContext = function (scope) {
			this.context = scope;
		};

		/**
		 * Handle a request from the native layer,
		 * this is supposed to be called directly from native code
		 *
		 * @public
		 *
		 * @param {String} data A JSON-encoded string containing the parameters
		 * associated to the request
		 */
		PontoDispatcher.prototype.request = function (data) {
			var params = new RequestParams(data),
				scope = this.context,
				target = scope[params.target];

			if (target) {
				dispatchRequest(scope, target, params);
			} else if (amd && params.target) {
				require([params.target], function (target) {
					dispatchRequest(scope, target, params);
				});
			}
		};

		/**
		 * Handle a response from the native layer,
		 * this is supposed to be called directly from native code
		 *
		 * @public
		 *
		 * @param {String} data A JSON-encoded string containing the parameters
		 * associated to the response
		 */
		PontoDispatcher.prototype.response = function (data) {
			var params = new ResponseParams(data);
			dispatchResponse(params);
		};

		/**
		 * Makes a request to the native layer
		 *
		 * @public
		 *
		 * @param {String} target The target native class
		 * @param {String} method The method to call
		 * @param {Object} params [OPTIONAL] An hash contaning the parameters to pass to the method
		 * @param {Function} completeCallback [OPTIONAL] The callback to invoke on completion
		 * @param {Function} errorCallback [OPTIONAL] The callback to invoke in case of error
		 * @param {Bool} async Indicates of the other side will make async operation and respond
		 * manually
		 */
		PontoDispatcher.prototype.invoke = function (target, method, params, completeCallback, errorCallback, async) {
			var callbackId;

			if (typeof (completeCallback || errorCallback) === 'function') {
				callbackId = Math.random().toString().substr(2);
				callbacks[callbackId] = {
					complete: completeCallback,
					error: errorCallback
				};
			}

			protocol.request(this.context, target, method, params, callbackId, async);
		};

		exports = dispatcher;
		exports.RESPONSE_COMPLETE = RESPONSE_COMPLETE;
		exports.RESPONSE_ERROR = RESPONSE_ERROR;
		exports.TARGET_NATIVE = TARGET_NATIVE;
		exports.TARGET_IFRAME = TARGET_IFRAME;
		exports.TARGET_IFRAME_PARENT = TARGET_IFRAME_PARENT;
		exports.PontoDispatcher = PontoDispatcher;
		exports.PontoBaseHandler = PontoBaseHandler;
		exports.setTarget = setTarget;

		return exports;
	}

	//check for AMD availability
	if (typeof define !== 'undefined' && define.amd) {
		amd = true;
	}

	//make module available in the global scope
	context.Ponto = ponto();

	//if AMD available then register also as a module
	//to allow easy usage in other AMD modules
	if (amd) {
		amd = true;
		define('ponto', context.Ponto);
	}
}(this));

File: android/ponto/src/com/wikia/ponto/Ponto.java
/**
 * @see  http://github.com/Wikia/ponto
 * @author Artur Klajnerok <arturk@wikia-inc.com>
 */

package com.wikia.ponto;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import org.json.JSONException;
import org.json.JSONObject;

import android.util.Log;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;

/**
 * TODO description
 */
public class Ponto {

    public static final String TAG = "Ponto";

    /**
     * Package that contains classes which can be invoked by WebView JavaScript
     */
    private String mClassPackage;

    /**
     * WebView that will communicate with native code
     */
    private WebView mWebView;

    /**
     * Callbacks to get WebView response for a request from the native code
     */
    private Map<String, RequestCallback> mCallbacks;

    public Ponto(WebView webView, String classPackage) {
        mCallbacks = new HashMap<String, RequestCallback>();
        mClassPackage = classPackage;
        mWebView = webView;
        mWebView.getSettings().setJavaScriptEnabled(true);
        mWebView.addJavascriptInterface(new PontoProtocol(), PontoProtocol.TAG);
    }

    /**
     * Makes a request to the WebView JavaScript
     * 
     * @param className The class name to instantiate.
     * @param methodName The method name to invoke.
     * @param params The parameters that will be passed to invoked method.
     * @param callback The callback for this request
     */
    public void invoke(String className, String methodName, String params, RequestCallback callback) {
        String callbackId = UUID.randomUUID().toString();

        if (callbackId != null && callback != null) {
            mCallbacks.put(callbackId, callback);
        }

        StringBuilder requestString = new StringBuilder();
        requestString.append("javascript:Ponto.request('{")
                .append("\"target\": \"").append(className).append("\", ")
                .append("\"method\": \"").append(methodName).append("\", ")
                .append("\"params\": ").append(params).append(", ")
                .append("\"callbackId\": \"").append(callbackId).append("\"")
                .append("}');");
        mWebView.loadUrl(requestString.toString());
    }

    /**
     * Communication protocol that should be used as JavascriptInterface
     */
    private class PontoProtocol {

        public static final String TAG = "PontoProtocol";

        /**
         * Represents a completed request
         */
        private static final int RESPONSE_COMPLETE = 0;

        /**
         * Represents a failed request with errors
         */
        private static final int RESPONSE_ERROR = 1;

        /**
         * The value of String that is null
         */
        private static final String NULL_STRING = "null";

        /**
         * The value of String that is undefined
         */
        private static final String UNDEFINED_STRING = "undefined";

        /**
         * Key for message response parameter
         */
        private static final String KEY_MESSAGE = "message";

        /**
         * Called by the web layer to perform a request on native layer
         * 
         * @param execContext The JavaScript context.
         * @param className The class name to instantiate.
         * @param methodName The method name to invoke.
         * @param params The parameters that will be passed to invoked method.
         * @param callbackId The id of web callback.
         * @param async 
         */
        @JavascriptInterface
        public void request(String execContext, String className, String methodName, String params, 
                String callbackId, String async) {

            int responseType = RESPONSE_ERROR;
            JSONObject responseParams = new JSONObject();
            try {
                Class<?> cls = Class.forName(mClassPackage + className);
                Constructor<?> constructor = cls.getConstructor(android.content.Context.class);
                Object object = constructor.newInstance(mWebView.getContext());
                if (params != null && !params.equalsIgnoreCase(NULL_STRING)) {
                    Method method = cls.getDeclaredMethod(methodName, String.class);
                    method.invoke(object, params);
                } else {
                    Method method = cls.getDeclaredMethod(methodName);
                    method.invoke(object);
                }
                responseType = RESPONSE_COMPLETE;
            } catch (InstantiationException e) {
                Log.e(TAG, "InstantiationException while executing ponto request", e);
            } catch (IllegalAccessException e) {
                Log.e(TAG, "IllegalAccessException while executing ponto request", e);
            } catch (ClassNotFoundException e) {
                Log.e(TAG, "ClassNotFoundException while executing ponto request", e);
                responseParams = getClassNotFoundParams(className);
            } catch (NoSuchMethodException e) {
                Log.e(TAG, "NoSuchMethodException while executing ponto request", e);
                responseParams = getNoSuchMethodParams(methodName);
            } catch (IllegalArgumentException e) {
                Log.e(TAG, "IllegalArgumentException while executing ponto request", e);
            } catch (InvocationTargetException e) {
                Log.e(TAG, "InvocationTargetException while executing ponto request", e);
            }

            if (callbackId != null && !callbackId.equalsIgnoreCase(UNDEFINED_STRING)) {
                javascriptCallback(callbackId, responseType, responseParams.toString());
            }
        }

        /**
         * Called by the web layer when answering a request from native layer
         * 
         * @param execContext The JavaScript context.
         * @param callbackId The id of callback that should be executed.
         * @param params The parameters associated with the response.
         */
        @JavascriptInterface
        public void response(String execContext, String callbackId, String params) {

            int responseType = getResponeTypeFromParams(params);
            if (callbackId != null && !callbackId.equalsIgnoreCase(UNDEFINED_STRING) &&
                    mCallbacks.containsKey(callbackId)) {

                RequestCallback callback = mCallbacks.get(callbackId);

                switch (responseType) {
                case RESPONSE_COMPLETE:
                    callback.onSuccess();
                    break;
                case RESPONSE_ERROR:
                default:
                    callback.onError();
                    break;
                }
                mCallbacks.remove(callbackId);
            }
        }

        /**
         * Makes a response to web layer
         * 
         * @param callbackId The id of callback that should be executed.
         * @param type The response type complete/error
         * @param params The parameters associated with the response.
         */
        private void javascriptCallback(final String callbackId, final int type, final String params) {
            mWebView.post(new Runnable() {
                @Override
                public void run() {
                    StringBuilder responseString = new StringBuilder();
                    responseString.append("javascript:Ponto.response('{")
                            .append("\"type\": ").append(type).append(", ")
                            .append("\"params\": ").append(params).append(", ")
                            .append("\"callbackId\": \"").append(callbackId).append("\"")
                            .append("}');");
                    mWebView.loadUrl(responseString.toString());
                }
            });
        }

        private int getResponeTypeFromParams(String params) {
            JSONObject responseParams;
            int type = RESPONSE_COMPLETE;
            try {
                responseParams = new JSONObject(params);
                if (responseParams != null) {
                    type = responseParams.optInt("type");
                }
            } catch (JSONException e) {
                Log.e(TAG, "JSONException while parsing messaging data", e);
            }
            return type;
        }

        private JSONObject getClassNotFoundParams(String className) {
            Map<String, String> paramsMap = new HashMap<String, String>();
            paramsMap.put(KEY_MESSAGE, "Class not found");
            paramsMap.put("className", className);
            return new JSONObject(paramsMap);
        }

        private JSONObject getNoSuchMethodParams(String methodName) {
            Map<String, String> paramsMap = new HashMap<String, String>();
            paramsMap.put(KEY_MESSAGE, "Method not found");
            paramsMap.put("methodName", methodName);
            return new JSONObject(paramsMap);
        }
    }

    /**
     * Interface definition for a callbacks that are invoked 
     * when WebView responds for a request from the native code.
     */
    public interface RequestCallback {

        /**
         * Callback method to be invoked when request was successful
         */
        public void onSuccess();

        /**
         * Callback method to be invoked when request failed
         */
        public void onError();
    }
}